### PR TITLE
Feat/merge microbial into samples

### PR DESCRIPTION
**Pull request by patrikgrenfeldt**
This PR merges the microbial orders/samples into families/samples to minimise the overhead of developing for both when they are so similar.

**How to prepare for test**:
- [x] ssh to clinical-db
- [x] install on stage of clinical-db:
`bash servers/resources/clinical-db.sclifelab.se/update-cgweb-stage.sh [THIS-BRANCH-NAME]`

**How to test**:
see servers PR

**Review:**
- [x] code approved by MM
- [x] tests executed by PG
- [x] "Merge and deploy" approved by PG
Thanks for filling in who performed the code review and the test!

This [version](https://semver.org/) is a:
- [ ] **MAJOR** - when you make incompatible API changes
- [x] **MINOR** - when you add functionality in a backwards compatible manner
- [ ] **PATCH** - when you make backwards compatible bug fixes or documentation/instructions

**To be deployed together**: 
- [ ] https://github.com/Clinical-Genomics/cg/pull/739
- [ ] https://github.com/Clinical-Genomics/cgweb/pull/218
- [ ] https://github.com/Clinical-Genomics/servers/pull/400

### CODE REVIEW

File: cg/cli/workflow/microsalt/base.py [hunk 1]
```diff
@@ -10,149 +10,117 @@
...
+
+    api = context.obj["api"]
+    sample_objs = api.get_samples(ticket, sample_id)
 
     if not sample_objs:
-        LOG.error("provide order and/or sample")
+        LOG.error("provide ticket and/or sample")
         context.abort()
 
     for sample_obj in sample_objs:
```
Comment (moonso) [inline cg/cli/workflow/microsalt/base.py:59]: Suggest to put in own function `link_microbial_samples` for easier testing

File: cg/cli/workflow/microsalt/base.py [hunk 2]
```diff
@@ -10,149 +10,117 @@
...
         LOG.info("%s: link FASTQ files", sample_obj.internal_id)
-        context.obj["api"].link_sample(
+        if dry_run:
+            continue
+        api.link_sample(
```
Comment (moonso) [inline cg/cli/workflow/microsalt/base.py:63]: Here for example a FastqHandler is instantiated for every sample in the loop. Suggest to do:
```python
def link_microbial_samples(samples: List[Sample], ticket: str, fastq_handler: FastqHandler, api: SomeAPI):
    """Links a list of microbial samples"""
    ...
```

File: cg/cli/workflow/microsalt/base.py [hunk 3]
```diff
@@ -10,149 +10,117 @@
...
-            LOG.error("Samples %s not found in %s ", sample_id, order_id)
-            context.abort()
-        sample_objs = [
-            sample_obj
-            for sample_obj in microbial_order_obj.microbial_samples
-            if sample_obj.internal_id == sample_id
-        ]
-    else:
-        LOG.error("provide order and/or sample")
+
```
Comment (moonso) [inline cg/cli/workflow/microsalt/base.py:77]: Really nice job on the code reduction/refactoring in these CLI commands @patrikgrenfeldt !

File: cg/cli/workflow/microsalt/base.py [hunk 4]
```diff
@@ -10,149 +10,117 @@
...
+    if not ticket and not sample_id:
```
Comment (moonso) [inline cg/cli/workflow/microsalt/base.py:78]:
```suggestion
    if not (ticket or sample_id):
```

File: cg/cli/workflow/microsalt/base.py [hunk 5]
```diff
@@ -10,149 +10,117 @@
...
+        LOG.error("Provide ticket and/or sample")
+        context.abort()
+
+    sample_objs = context.obj["api"].get_samples(ticket, sample_id)
```
Comment (moonso) [inline cg/cli/workflow/microsalt/base.py:82]: Again I think "api" is a bit vague

File: cg/cli/workflow/microsalt/base.py [hunk 6]
```diff
@@ -10,149 +10,117 @@
...
 
-    filename = order_id if order_id else sample_id
-    outfilename = Path(context.obj["usalt"]["queries_path"]) / filename
-    outfilename = outfilename.with_suffix(".json")
-    if dry:
+    filename = str(ticket) if ticket else sample_id
+    outfilename = (Path(context.obj["usalt"]["queries_path"]) / filename).with_suffix(".json")
+    if dry_run:
         print(json.dumps(parameters, indent=4, sort_keys=True))
     else:
```
Comment (moonso) [inline cg/cli/workflow/microsalt/base.py:96]:
```suggestion
    if dry_run:
        click.echo(json.dumps(parameters, indent=4, sort_keys=True))
        return
```
Then flatten below

File: cg/cli/workflow/microsalt/base.py [hunk 7]
```diff
@@ -10,149 +10,117 @@
...
+    "-c",
+    "--config-case",
+    "config_case_path",
+    required=False,
+    help="optionally change the config-case",
+)
+@click.argument("ticket")
 @click.pass_context
-def run(context, dry, config_case, order_id):
+def run(context, dry_run, config_case_path, ticket):
```
Comment (moonso) [inline cg/cli/workflow/microsalt/base.py:112]: Type hints here?

File: cg/cli/workflow/microsalt/base.py [hunk 8]
```diff
@@ -10,149 +10,117 @@
...
-    if not config_case:
+    if not config_case_path:
         queries_path = Path(context.obj["usalt"]["queries_path"])
-        config_case_path = queries_path / order_id
-        config_case_path = config_case_path.with_suffix(".json")
+        config_case_path = (queries_path / ticket).with_suffix(".json")
 
     command.extend(["--parameters", str(config_case_path)])
-    if dry:
+    if dry_run:
```
Comment (moonso) [inline cg/cli/workflow/microsalt/base.py:123]: Flatten and use click.echo instead of print

File: cg/meta/invoice.py [hunk 9]
```diff
@@ -146,12 +146,8 @@ def prepare_record(self, costcenter: str, discount: int, record):
...
         )
 
-        order = record.microbial_order.id if self.record_type == "Microbial" else record.order
-        ticket_number = (
-            record.microbial_order.ticket_number
-            if self.record_type == "Microbial"
-            else record.ticket_number
-        )
+        order = record.order
+        ticket_number = record.ticket_number
```
Comment (moonso) [inline cg/meta/invoice.py:150]: 😸 this is nice!

File: cg/meta/microsalt/lims.py [hunk 10]
```diff
@@ -30,13 +30,23 @@ def get_lims_comment(self, sample_id: str) -> str:
 
         return ""
 
-    def get_organism(self, sample_obj: Sample) -> str:
+    def get_lims_organism(self, sample_id: str) -> str:
```
Comment (moonso) [inline cg/meta/microsalt/lims.py:33]: Well you know what I think... Why is this function in two places?

File: cg/meta/orders/status.py [hunk 11]
```diff
@@ -391,28 +367,37 @@ def store_microbial_order(
                     )
                     self.status.add_commit(organism)
 
-                new_sample = self.status.add_microbial_sample(
+                if comment:
+                    sample_comment = f"Order comment: {comment}"
```
Comment (moonso) [inline cg/meta/orders/status.py:371]: So if I understand we store two informations in one field here as a string. Would it not be preferable to have different fields for this information?

**Review by moonso — CHANGES_REQUESTED** [review]
Great job @patrikgrenfeldt ! Looks like you could remove a lot of almost redundant code with this PR. There are however some things that should be up for discussion, I suggest we talk about that on the meeting on Monday 14/9.

File: cg/cli/workflow/get_links.py [hunk 12]
```diff
@@ -10,58 +10,25 @@
...
-                "Could not find sample %s. Did you intend %s as a case-id?",
-                sample_id,
-                sample_id,
-            )
-            raise click.Abort
-
-        if not sample_obj.links:
-            LOG.error("Could not find links for sample %s", sample_id)
-            raise click.Abort
+    if not case_id and not sample_id and not ticket:
```
[the remark this answers is not in the document]
Comment (patrikgrenfeldt) [reply cg/cli/workflow/get_links.py]: rewritten, better?

File: cg/cli/workflow/microsalt/base.py [hunk 13]
```diff
@@ -10,149 +10,117 @@
...
-        context.obj,
-        hk_api=hk_api,
-        lims_api=lims_api,
-        case_tags=PROJECT_TAGS,
-        sample_tags=SAMPLE_TAGS,
-    )
     context.obj["api"] = AnalysisAPI(db=context.obj["db"], hk_api=hk_api, lims_api=lims_api)
     context.obj["lims_microsalt_api"] = LimsMicrosaltAPI(lims=lims_api)
 
     if context.invoked_subcommand is None:
```
[the remark this answers is not in the document]
Comment (patrikgrenfeldt) [reply cg/cli/workflow/microsalt/base.py]: Flattened

File: cg/meta/invoice.py (hunk 9 above)
Comment (patrikgrenfeldt) [reply cg/meta/invoice.py:150]: thanks 😄

File: cg/cli/workflow/microsalt/base.py (hunk 8 above)
Comment (patrikgrenfeldt) [reply cg/cli/workflow/microsalt/base.py:130]: flattened

File: cg/meta/orders/status.py (hunk 11 above)
Comment (patrikgrenfeldt) [reply cg/meta/orders/status.py:371]: sure, how?

File: cg/meta/microsalt/lims.py (hunk 10 above)
Comment (patrikgrenfeldt) [reply cg/meta/microsalt/lims.py:33]: are you saying it is also in the lims-api?

File: cg/cli/workflow/microsalt/base.py (hunk 7 above)
Comment (patrikgrenfeldt) [reply cg/cli/workflow/microsalt/base.py]: fixed

File: cg/cli/workflow/microsalt/base.py [hunk 14]
```diff
@@ -10,149 +10,117 @@
...
     lims_api = lims.LimsAPI(context.obj)
-    deliver = DeliverAPI(
-        context.obj,
-        hk_api=hk_api,
-        lims_api=lims_api,
-        case_tags=PROJECT_TAGS,
-        sample_tags=SAMPLE_TAGS,
-    )
     context.obj["api"] = AnalysisAPI(db=context.obj["db"], hk_api=hk_api, lims_api=lims_api)
     context.obj["lims_microsalt_api"] = LimsMicrosaltAPI(lims=lims_api)
```
[the remark this answers is not in the document]
Comment (patrikgrenfeldt) [reply cg/cli/workflow/microsalt/base.py:32]: Are you suggesting it should be merged with cg/meta/workflow/microsalt.py?

[the remark this answers is not in the document]
Comment (moonso) [reply cg/cli/workflow/microsalt/base.py:31]: Maybe that or the LIMS api

File: cg/meta/orders/status.py (hunk 11 above)
Comment (moonso) [reply cg/meta/orders/status.py:371]: One columns for order comment and one for sample comment

File: requirements-dev.txt [hunk 15]
```diff
@@ -3,5 +3,5 @@ pytest>=5.2     # due to pytest-flask, pytest-cov
 mock
 pytest-mock
 snapshottest
-
+deprecated
```
Comment (moonso) [inline requirements-dev.txt:6]: Did we have a dependency called deprecated or what is this?

File: cg/cli/workflow/get_links.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (moonso) [reply cg/cli/workflow/get_links.py]: I can not see that it has changed...

File: requirements-dev.txt (hunk 15 above)
Comment (patrikgrenfeldt) [reply requirements-dev.txt]: I removed the usage of the deprecated method and the usage of this library

File: cg/meta/orders/status.py (hunk 11 above)
Comment (patrikgrenfeldt) [reply cg/meta/orders/status.py:371]: removed the order comment completely

File: cg/meta/microsalt/lims.py (hunk 10 above)
Comment (patrikgrenfeldt) [reply cg/meta/microsalt/lims.py]: removed

File: cg/cli/workflow/microsalt/base.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (patrikgrenfeldt) [reply cg/cli/workflow/microsalt/base.py]: I merged them now

File: cg/cli/workflow/get_links.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (patrikgrenfeldt) [reply cg/cli/workflow/get_links.py]: maybe you have a "refresh" link in the "files" tab? That could be it I guess

File: cg/apps/lims/api.py [hunk 16]
```diff
@@ -381,3 +376,21 @@ def _find_twist_capture_kits(artifacts, udf_key):
             if artifact.udf.get(udf_key) is not None
         )
         return capture_kits
+
+    def get_sample_organism(self, lims_id):
```
[the remark this answers is not in the document]
Comment (patrikgrenfeldt) [reply cg/apps/lims/api.py]: it's back on sample now

File: cg/cli/workflow/get_links.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (patrikgrenfeldt) [reply cg/cli/workflow/get_links.py]: or just look [here](https://github.com/Clinical-Genomics/cg/pull/739/files#diff-314526873468d59f90f4d37ae0f5e7e9)

File: cg/apps/lims/api.py [hunk 17]
```diff
@@ -381,3 +376,12 @@ def _find_twist_capture_kits(artifacts, udf_key):
             if artifact.udf.get(udf_key) is not None
         )
         return capture_kits
+
+    def get_sample_comment(self, sample_id: str):
```
Comment (moonso) [inline cg/apps/lims/api.py:380]: Could you add return types for these? like:
```suggestion
    def get_sample_comment(self, sample_id: str) -> str:
```
Or whatever it is returning

File: cg/apps/lims/api.py [hunk 18]
```diff
@@ -381,3 +376,12 @@ def _find_twist_capture_kits(artifacts, udf_key):
...
+        """Get the comment of the sample"""
+        lims_sample = self.sample(sample_id)
+        return lims_sample.get("comment")
+
+    def get_sample_project(self, sample_id: str):
```
Comment (moonso) [inline cg/apps/lims/api.py:385]: Returntype

File: cg/apps/microsalt/fastq.py [hunk 19]
```diff
@@ -36,12 +36,12 @@ def __init__(self, config):
         super().__init__(config)
         self.root_dir = config["usalt"]["root"]
 
-    def link(self, case: str, sample: str, files: List):
+    def link(self, ticket: int, sample: str, files: List):
```
Comment (moonso) [inline cg/apps/microsalt/fastq.py:39]: List of what? Could you also add return type? Suggestion
```suggestion
    def link(self, ticket: int, sample: str, files: List[Path]) -> None:
```
I think we should aim for a situation where we can run mypy checks on all our code

File: cg/cli/workflow/get_links.py [hunk 20]
```diff
@@ -10,62 +10,25 @@
...
-                "Could not find sample %s. Did you intend %s as a case-id?",
-                sample_id,
-                sample_id,
-            )
-            raise click.Abort
-
-        if not sample_obj.links:
-            LOG.error("Could not find links for sample %s", sample_id)
-            raise click.Abort
+    if not any([case_id, sample_id, ticket]):
```
Comment (moonso) [inline cg/cli/workflow/get_links.py:24]: 👍

File: cg/cli/workflow/microsalt/base.py [hunk 21]
```diff
@@ -10,149 +10,117 @@
...
-        # link all samples in a case
-        sample_objs = context.obj["db"].microbial_order(order_id).microbial_samples
-    elif sample_id and (order_id is None):
-        # link sample in all its families
-        sample_objs = [context.obj["db"].microbial_sample(sample_id)]
-    elif sample_id and order_id:
-        # link only one sample in a case
-        sample_objs = [context.obj["db"].microbial_sample(sample_id)]
+
+    api = context.obj["api"]
```
[the remark this answers is not in the document]
Comment (moonso) [reply cg/cli/workflow/microsalt/base.py:54]: +1 on this to myself

File: cg/cli/workflow/microsalt/base.py (hunk 2 above)
Comment (moonso) [reply cg/cli/workflow/microsalt/base.py:65]: What did you think of this @patrikgrenfeldt ?

File: cg/meta/workflow/microsalt.py [hunk 22]
```diff
@@ -1,49 +1,49 @@
+""" API to manage Microsalt Analyses
```
Comment (moonso) [inline cg/meta/workflow/microsalt.py:1]: 👍

File: cg/meta/workflow/microsalt.py [hunk 23]
```diff
@@ -1,49 +1,49 @@
...
-        logger=logging.getLogger(__name__),
     ):
         self.db = db
         self.hk = hk_api
         self.lims = lims_api
-        self.yaml_loader = yaml_loader
-        self.pather = path_api
-        self.LOG = logger
 
     def check(self, family_obj: models.Family):
```
Comment (moonso) [inline cg/meta/workflow/microsalt.py:35]: Returntype?

File: cg/meta/workflow/microsalt.py [hunk 24]
```diff
@@ -103,7 +100,7 @@ def fastq_header(line):
 
         return rs
 
-    def link_sample(self, fastq_handler: BaseFastqHandler, sample: str, case: str):
+    def link_sample(self, fastq_handler: FastqHandler, sample: str, ticket: int):
```
Comment (moonso) [inline cg/meta/workflow/microsalt.py:103]: Returntype. `None` is also an important return type

File: cg/meta/workflow/microsalt.py [hunk 25]
```diff
@@ -127,4 +124,80 @@ def link_sample(self, fastq_handler: BaseFastqHandler, sample: str, case: str):
...
+            if reference == "NC_017960.1":
+                organism = "Enterococcus faecium"
+            elif reference == "NC_004668.1":
+                organism = "Enterococcus faecalis"
+            elif has_comment:
+                organism = comment
+
+        return organism
+
+    def get_parameters(self, sample_obj: Sample):
```
Comment (moonso) [inline cg/meta/workflow/microsalt.py:171]:
```suggestion
    def get_parameters(self, sample_obj: Sample) -> Dict[str,str]:
```

**Review by moonso — CHANGES_REQUESTED** [review]
Nice, very large PR. I did most of the review now. Might need to return to it and perhaps @Mropat can have a look as well

File: cg/meta/workflow/microsalt.py (hunk 23 above)
Comment (patrikgrenfeldt) [reply cg/meta/workflow/microsalt.py]: added

File: cg/meta/workflow/microsalt.py (hunk 24 above)
Comment (patrikgrenfeldt) [reply cg/meta/workflow/microsalt.py]: added

File: cg/meta/workflow/microsalt.py (hunk 25 above)
Comment (patrikgrenfeldt) [reply cg/meta/workflow/microsalt.py]: added

**Comment by patrikgrenfeldt** [comment]
@moonso I did not understand your question, 
image

Could you clarify your question?

**Comment by moonso** [comment]
> Could you clarify your question?

Could we keep the discussions on the actual comments in question?

Anyway I gave a suggestion that you put this piece of code in its own function (link_microbial_fastsq_files)so you not instantiate the fastq_api in every loop.

File: cg/apps/lims/api.py (hunk 18 above)
Comment (patrikgrenfeldt) [reply cg/apps/lims/api.py]: added

**Comment by patrikgrenfeldt** [comment]
> Could we keep the discussions on the actual comments in question?

Yes, preferably. But there was no field to enter text in, see image:
image

File: cg/apps/microsalt/fastq.py (hunk 19 above)
Comment (patrikgrenfeldt) [reply cg/apps/microsalt/fastq.py]: fixed

**Review by moonso — APPROVED** [review]

**Comment by patrikgrenfeldt** [comment]
bumped

**Comment by patrikgrenfeldt** [comment]
deployed on clinical-api: 



deplyed on Hasta:

**Pull request merged**